active/0003-named-template-blocks.md

```diff
@@ -0,0 +1,126 @@
+
+Here is what that might look like,
+
+```
+{{#ember-table}}
+  {{^header as |name|}}
+  	{{name}}
+  {{^cell as |value|}}
+  	{{value}}
+  {{^}}
+  Nothing to show.
+{{/ember-table}}
+```
+
+Specifying a named block inside of the component's block template would override default implementation of that block inside of the component's layout.
+
+# Detailed design
+
+## Expand ^ syntax
+
+Currently, Handlebars implements `^` which means `else`. Infact, `else` is aliased to `^`. We would expand this syntax to allow named portions of template. 
+
+Here is an example of what a template with this syntax might look like.
+
+```
+{{#table-component items as |item|}}
+	// default block
+	{{^header}}
+	// header content
+	{{^footer}}
+	// footer content
+	{{^}}
+	// empty view
+{{/table-component}}
+```
+
+`{{^name}}` serve as dividers of the block template.
+
+## Refactor `{{else}}` helper
+
+One possible implementation would be to make `{{^}}` implementation and allow the name of the helper to be specified. 
+
+Handlebars has several built in blocks that are availble on helper's `options` argument,  namely `options.fn` & `options.inverse`. These would be changed to `options.blocks.default` & `options.blocks.inverse` respectively. Every other named template block would be available on `options.blocks` hash. The above example would have `options.blocks.header` & `options.blocks.footer` in addition to it's default blocks.
+
+The component hook will append named blocks onto the template as it does currently with default block.
+
+## Block are available in the layout
+
+The component must be able to determine programmatically if it should consume it's default block or use the passed in named block. If we consider the above example, then `table-component`'s layout might look something like this.
+
+```
+{{#if blocks.header}}
+	{{yield-to 'header' headerContent}}
+{{else}}
+	<thead>
+		{{#each headerContent as |name|}}
+			<th>{{name}}</th>
+		{{/each}}
+	</thead>
+{{/if}}
+```
+
+`blocks` keyword becomes a reserved keyword and will throw a warning when the component defines a *blocks* property. We already have a reserved `hasBlock` keyword in the template, so this will not be a far stretch.
+
+## Named blocks have block params
+
+Named blocks will have block params and will receive values with a new `{{yield-to}}` helper. `{{yield-to}}` will take name of a block as a first parameter and yield properties as `{{yield}}` does. This will allow the component to expose template friendly values to be used in the named blocks.
+
+```{{yield-to 'header' headerContent}}``` will yield `headerContent` to `header` block.
+
+# Drawbacks
+
+Why should we *not* do this?
+
+# Alternatives
```

Comment:
`yield-to` has no idea what to do if there are multiple `named` template blocks, imagine the following.

```hbs
<my-form as |f|>

  <f.input> </f.input>
  <f.input> </f.input>
  <f.input> </f.input>

  <f.save> Save</f.save>
</my-component>
```

or a table example:

```js
<my-table as |t|>
  <t.column> my first column </t.column>
  <t.column> my second column </t.column>
  <t.column> my thirds column </t.column>
</my-table>

```